core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java

```diff
@@ -169,35 +173,107 @@ protected PlanNodeStatsEstimate visitLogicalExpression(LogicalExpression node, V
 
         private PlanNodeStatsEstimate estimateLogicalAnd(List<Expression> terms)
         {
-            // first try to estimate in the fair way
-            PlanNodeStatsEstimate estimate = process(terms.get(0));
-            if (!estimate.isOutputRowCountUnknown()) {
-                for (int i = 1; i < terms.size(); i++) {
-                    estimate = new FilterExpressionStatsCalculatingVisitor(estimate, session, types).process(terms.get(i));
-
-                    if (estimate.isOutputRowCountUnknown()) {
-                        break;
-                    }
-                }
+            List<PlanNodeStatsEstimate> estimates = estimateExpressions(terms);
+            // Sort known estimates by selectivity
+            List<PlanNodeStatsEstimate> knownSortedEstimates = estimates.stream()
+                    .filter(estimate -> !estimate.isOutputRowCountUnknown())
+                    .sorted(Comparator.comparingDouble(PlanNodeStatsEstimate::getOutputRowCount))
+                    .collect(toImmutableList());
 
-                if (!estimate.isOutputRowCountUnknown()) {
-                    return estimate;
+            if (knownSortedEstimates.isEmpty()) {
+                return PlanNodeStatsEstimate.unknown();
+            }
+
+            PlanNodeStatsEstimate combinedEstimate = knownSortedEstimates.get(0);
+            double combinedSelectivity = combinedEstimate.getOutputRowCount() / input.getOutputRowCount();
+            double combinedScalingFactor = 1.0;
+            double scalingFactor = getFilterConjunctionSelectivityWeight(session);
+            // For scalingFactor = 0.75 and terms t1, t2, t3
+            // Combined selectivity = (t1 selectivity) * ((t2 selectivity) ^ 0.75) * ((t3 selectivity) ^ 0.5625)
+            // scalingFactor = 1 implies the terms are assumed to have no correlation and their selectivities are multiplied without scaling.
+            // scalingFactor = 0 implies the terms are assumed to be fully correlated and only the most selective term drives the selectivity.
+            for (int i = 1; i < knownSortedEstimates.size(); i++) {
+                PlanNodeStatsEstimate termEstimate = knownSortedEstimates.get(i);
+                combinedScalingFactor *= scalingFactor;
+                combinedSelectivity *= Math.pow(termEstimate.getOutputRowCount() / input.getOutputRowCount(), combinedScalingFactor);
+                // Scale output row count by combinedSelectivityFactor for every additional term
+                double outputRowCount = Math.max(input.getOutputRowCount() * combinedSelectivity, 1);
+                combinedEstimate = combineEstimatesAnd(combinedEstimate, termEstimate, outputRowCount, types);
+            }
+            boolean unestimatedTermFound = estimates.stream().anyMatch(PlanNodeStatsEstimate::isOutputRowCountUnknown);
+            return combinedEstimate.mapOutputRowCount(rowCount -> unestimatedTermFound ? rowCount * UNKNOWN_FILTER_COEFFICIENT : rowCount);
+        }
+
```

Comment:
Let's extract another method, which is `extractCorrelatedExpressions` which would return, `Set<Set<Expression>>`